content/data/scripts/ui_topics.lua

```diff
@@ -21,24 +22,47 @@ local function weaponStats(weaponClass)
 	local baseDamage = weaponClass.Damage
 	if weaponClass.OuterRadius > 0 then
 		-- This weapon has a shockwave, which gives it additional damage on a direct hit.
-		-- TODO: Most weapons have shockwave damage equal to their base damage, but not all.
-		local bonusDamage = baseDamage
+		-- Added formula to calculate shockwave damage -- WW
+		if weaponClass.ShockwaveDamage and weaponClass.ShockwaveDamage > 0 then
+			bonusDamage = weaponClass.ShockwaveDamage
+		elseif weaponClass.ShockwaveDamage == 0 or weaponClass.ShockwaveDamage == nil then
+			bonusDamage = baseDamage
+		end
 		baseDamage = baseDamage + bonusDamage
 	end
+
 	local velocity = weaponClass.Speed
-	local range = velocity * weaponClass.LifeMax
-	local rof = math.floor(100 / weaponClass.FireWait) / 100
+	local rof = utils.round(1 / weaponClass.FireWait, 2)
+	local range = math.min(weaponClass.Range, (velocity * weaponClass.LifeMax))
+
+	-- new code to calculate volley size -- WW
+	local isSwarmer, swarmcount, swarmwait = weaponClass:getSwarmInfo()
+	local isCorkscrew, cscrewcount, cscrewwait, cscrewrotate, cscrewtwist = weaponClass:getCorkscrewInfo()
+	local swarmproxy = math.max(swarmcount, 1) --swarmcount returns -1 if invalid
```

Comment:
Many of the above variables should be camelcase which will make the code easier to read. There's also not much need to abbreviate variables here and it would be more readable to just spell out corkscrew. Additionally it looks like you are setting variables that don't ever get used below. You can just leave those out.

`local isCorkscrew, corkscrewCount = weaponClass:getCorkscrewInfo()`

Reply:
Done all of these.